xray/commands/audit/scan.go

```diff
@@ -0,0 +1,317 @@
+package audit
+
+import (
+	"encoding/json"
+	"os"
+	"regexp"
+
+	"github.com/jfrog/gofrog/io"
+	"github.com/jfrog/gofrog/parallel"
+	"github.com/jfrog/jfrog-cli-core/common/spec"
+	"github.com/jfrog/jfrog-cli-core/utils/config"
+	"github.com/jfrog/jfrog-cli-core/utils/coreutils"
+	"github.com/jfrog/jfrog-cli-core/xray/commands"
+	xrutils "github.com/jfrog/jfrog-cli-core/xray/utils"
+	"github.com/jfrog/jfrog-client-go/artifactory/services/fspatterns"
+	clientutils "github.com/jfrog/jfrog-client-go/utils"
+	"github.com/jfrog/jfrog-client-go/utils/errorutils"
+	"github.com/jfrog/jfrog-client-go/utils/io/fileutils"
+	"github.com/jfrog/jfrog-client-go/utils/log"
+	"github.com/jfrog/jfrog-client-go/xray/services"
+)
+
+const (
+	IndexerExecutionName = "indexer-app"
+	IndexingCommand      = "graph"
+)
+
+type FileContext func(string) parallel.TaskFunc
+type indexFileHandlerFunc func(file string)
+
+type ScanCommand struct {
+	serverDetails *config.ServerDetails
+	spec          *spec.SpecFiles
+	threads       int
+	// The location on the local file system of the downloaded Xray's indexer.
+	indexerPath            string
+	printResults           bool
+	projectKey             string
+	watches                []string
+	includeVulnerabilities bool
+	includeLincenses       bool
+	scanPassed             bool
+}
+
+func (scanCmd *ScanCommand) SetThreads(threads int) *ScanCommand {
+	scanCmd.threads = threads
+	return scanCmd
+}
+
+func (scanCmd *ScanCommand) SetPrintResults(print bool) *ScanCommand {
+	scanCmd.printResults = print
+	return scanCmd
+}
+
+func (scanCmd *ScanCommand) SetServerDetails(server *config.ServerDetails) *ScanCommand {
+	scanCmd.serverDetails = server
+	return scanCmd
+}
+
+func (scanCmd *ScanCommand) SetSpec(spec *spec.SpecFiles) *ScanCommand {
+	scanCmd.spec = spec
+	return scanCmd
+}
+
+func (scanCmd *ScanCommand) SetProject(project string) *ScanCommand {
+	scanCmd.projectKey = project
+	return scanCmd
+}
+
+func (scanCmd *ScanCommand) SetWatches(watches []string) *ScanCommand {
+	scanCmd.watches = watches
+	return scanCmd
+}
+
+func (scanCmd *ScanCommand) SetIncludeVulnerabilities(include bool) *ScanCommand {
+	scanCmd.includeVulnerabilities = include
+	return scanCmd
+}
+
+func (scanCmd *ScanCommand) SetIncludeLincenses(include bool) *ScanCommand {
+	scanCmd.includeLincenses = include
+	return scanCmd
+}
+
+func (scanCmd *ScanCommand) ServerDetails() (*config.ServerDetails, error) {
+	return scanCmd.serverDetails, nil
+}
+
+func (scanCmd *ScanCommand) IsScanPassed() bool {
+	return scanCmd.scanPassed
+}
+
+func (scanCmd *ScanCommand) indexFile(filePath string) (*services.GraphNode, error) {
+	var indexerResults services.GraphNode
+	indexCmd := &coreutils.GeneralExecCmd{
+		ExecPath: scanCmd.indexerPath,
+		Command:  []string{IndexingCommand, filePath},
+	}
+	output, err := io.RunCmdOutput(indexCmd)
```

Comment:
Let's try to make the indexer fail, and see thay an error is indeed return in that case.